File: examples/exo/python/src/quick_start.py
from __future__ import annotations

from qnbot_sdk import Sample, Sdk
from qnbot_sdk.exo import ExoConfig, ExoTelemetry


def print_telemetry(sample: Sample[ExoTelemetry]) -> None:
    print(
        f"telemetry sequence={sample.sequence} "
        f"left_arm={sample.value.payload.left_arm.joint_positions_rad}"
    )


def main() -> None:
    sdk = Sdk(devices=(ExoConfig(),))
    exo = sdk.exo()
    device = exo.device()
    device.telemetry().subscribe(print_telemetry)

    exo.start()
    print("running; press Ctrl+C to stop", flush=True)
    exo.run_forever()


if __name__ == "__main__":
    main()

File: examples/exo/cpp/src/quick_start.cpp
#include "exo_example_support.hpp"

#include <cstdlib>
#include <exception>
#include <iostream>
#include <utility>

int main() {
    try {
        qnbot::SdkConfig config;
        config.devices = {qnbot::ExoConfig{}};

        qnbot::Sdk sdk(std::move(config));
        auto exo = sdk.exo();
        auto device = exo.device();
        auto telemetry = device.telemetry();
        const auto subscription = telemetry.subscribe(
            [](const qnbot::Sample<qnbot::ExoTelemetry>& sample) {
                std::cout << "telemetry sequence=" << sample.sequence
                          << " left_arm=";
                example::print_values(
                    sample.value.payload.left_arm.joint_positions_rad);
                std::cout << '\n';
            });

        exo.start();
        std::cout << "running; press Ctrl+C to stop" << std::endl;
        exo.run_forever();
        static_cast<void>(subscription);
        return EXIT_SUCCESS;
    } catch (const std::exception& error) {
        return example::report_error(error);
    }
}
